### PR TITLE
Deduplicate runner structs

### PR DESCRIPTION
**Pull request by tiziano88**
# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by tiziano88** [comment]
It turns out that the hierarchy nested structs can be flattened out to just two function calls to `run_example` and `run_function_example` with the appropriate argument types (which makes sense since there is only a single dispatch point conceptually: either run a classic example or a functions example, so we should expect a single dispatch based on this bit of information). I think there is still a corner case for examples with no applications (e.g. the tls_attestation one), but that may be sorted out later.

File: runner/src/examples.rs [hunk 1]
```diff
@@ -478,16 +256,73 @@ fn run_example(example: &ClassicExample) -> Step {
     // clients in the foreground.
     #[allow(clippy::collapsible_if)]
     let run_backend_server_clients: Step = if opt.run_server.unwrap_or(true) {
-        let run_server_clients = example.construct_example_server_run_step(run_clients);
-        example.construct_backend_run_steps(run_server_clients)
+        let run_server_clients = match example.applications.get(app_variant) {
```
Comment (rbehjati) [inline runner/src/examples.rs:259]: `run_example` and `run_functions_example` functions are now too long, too complex, and have too many responsibilities. This reduces readability and makes future modifications to the code very difficult, especially for someone who is new to this part of the code. I suggest refactoring the inlined code into the following functions:
`construct_example_server_run_step`, `construct_backend_run_steps`, `construct_application_build_steps` and similar functions for the `functions` version. 

This would show the overall functionality of `run_example` and `run_functions_example` much better, and would make it clear that these two functions are doing the same thing.

**Review by rbehjati — APPROVED** [review]

File: runner/src/examples.rs (hunk 1 above)
Comment (tiziano88) [reply runner/src/examples.rs:259]: You are right, let me try to move things around some more. Maybe I can go back to roughly the same functions that you had, but just make them separate top-level functions with explicit arguments, instead of methods defined on traits and structs (which resembles too much OOP to me and I find a bit hard to follow in the context of the rest of the Rust code which looks more functional). Will ping you if / when I get anywhere with that :)
